android-components/components/feature/search/src/main/java/mozilla/components/feature/search/SearchUseCases.kt

```diff
@@ -77,16 +81,21 @@ class SearchUseCases(
 
             val id = if (sessionId == null) {
                 // If no `sessionId` was passed in then create a new tab
-                tabsUseCases.addTab(searchUrl)
+                tabsUseCases.addTab(searchUrl, isSearch = true)
             } else {
                 // If we got a `sessionId` then try to find the tab and load the search URL in it
                 val existingTab = store.state.findTabOrCustomTab(sessionId)
                 if (existingTab != null) {
-                    sessionUseCases.loadUrl(searchUrl, sessionId = existingTab.id)
+                    store.dispatch(ContentAction.UpdateIsSearchAction(existingTab.id, true))
+                        .invokeOnCompletion @OptIn(DelicateCoroutinesApi::class) {
```

Comment:
What's interesting/challenging here is that we do this to prevent a race condition for the optimized case in `SessionUseCases`, which loads the URL directly on an existing engine session to prevent a context switch. However, with this, and I haven't considered this before, we now introduce another context switch, which will make search slower again in practice.

Did you actually reproduce a problem here in practice? Did you see a case where the state in the store wasn't yet updated without `invokeOnCompletion`? A bit unfortunate, that we have this context switch here now. If this is only theoretical, I could also file a follow and we remove the `invokeOnCompletion` for now. I need to think about this some more.

Reply:
Also, thanks @Alexandru2909 for working through this. It's a surprisingly tricky problem, but we're close to landing it.

Reply:
In the previous iteration this case was unintentionally taken care of, but after changing the implementation and a bit of testing I found the problem is reproducible and decided to add `invokeOnCompletion`.

Reply:
@Alexandru2909 OK, then I propose the following simplification and we can land it: https://github.com/csadilek/firefox-android/commit/71b4a5d4cc0d80a839783800f5a9d7063446ff1c

Actions are processed in order anyway, and this way we don't need the callback and context switch. This still means that this search use case will be slightly slower, but there' no way around that and still fixing this bug. So, I'd be OK with it.

Wdyt?

Reply:
Sounds good to me, added the change now

Reply:
👍 OK, let's land after the release branch is cut. There's one nit below about the ktlint-baseline.